### PR TITLE
Allows pre-loading reagents into the Vali system

### PR DESCRIPTION
**Pull request by Walarks**
## About The Pull Request
Allows loading up to 60u of reagents into the Vali system that you can pick to be automatically injected into you when you activate the suit. This is done through an action in the configurations radial menu.

**How to use:**
> You can't load pills into it.
 - When you are holding a reagent container filled with something, you can load the reagent into the vali's reagent container.
 - When you are holding an empty reagent container, you can empty the vali's reagent container into it.
 - When you aren't holding a reagent container, you can pick whether you would like for the suit's reagents to be automatically injected on activation or not.

## Why It's Good For The Game
Makes the system's reagent management less clunky to use.

## Changelog
:cl:
qol: allows loading reagents into the Vali system that are automatically injected on activation
/:cl:

### CODE REVIEW

**Comment by tgstation-server** [comment]
You currently have a negative Fix/Feature pull request delta of -37. Maintainers may close this PR at will. Fixing issues or improving the codebase will improve this score.

**Comment by Walarks** [comment]
I'm not sure how much of this is a feature or QoL.

File: code/datums/components/chem_booster.dm [hunk 1]
```diff
@@ -39,6 +40,10 @@
 	var/obj/item/connected_weapon
 	///When was the effect activated. Used to activate negative effects after a certain amount of use
 	var/processing_start = 0
+	///Internal reagent storage used to store and automatically inject reagents into the wearer
+	var/obj/item/reagent_containers/glass/beaker/meds_beaker = null
+	///Weather the contents on the meds_beaker will be injected into the wearer when the system is turned on
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:45]: How is the weather today?

File: code/datums/components/chem_booster.dm [hunk 2]
```diff
@@ -48,13 +53,15 @@
...
 	analyzer = new
+	meds_beaker = new /obj/item/reagent_containers/glass/beaker
 
 /datum/component/chem_booster/Destroy(force, silent)
 	QDEL_NULL(configure_action)
 	QDEL_NULL(power_action)
 	QDEL_NULL(scan_action)
 	QDEL_NULL(analyzer)
 	wearer = null
+	meds_beaker = null
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:64]: Not going to qdel_null?

File: code/datums/components/chem_booster.dm [hunk 3]
```diff
@@ -330,6 +345,60 @@
 	held_item.reagents.add_reagent(/datum/reagent/virilyth, amount)
 	extract(volume)
 
+///Fills an internal beaker that gets injected into the wearer on suit activation
+/datum/component/chem_booster/proc/load_up()
+	if(wearer.do_actions)
+		return
+
+	var/obj/item/held_item = wearer.get_held_item()
+	if(((!held_item || !istype(held_item, /obj/item/reagent_containers)) && !meds_beaker.reagents.total_volume) || istype(held_item, /obj/item/reagent_containers/pill))
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:354]: Istype null will always fail

File: code/datums/components/chem_booster.dm [hunk 4]
```diff
@@ -330,6 +345,60 @@
...
+		to_chat(wearer, "<span class='warning'>You need to be holding a liquid container to fill up the system's reagent storage.</span>")
+		return
+
+	if((!held_item || !istype(held_item, /obj/item/reagent_containers)) && meds_beaker.reagents.total_volume)
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:358]: Ditto

File: code/datums/components/chem_booster.dm [hunk 5]
```diff
@@ -330,6 +345,60 @@
...
+	show_meds_beaker_contents()
+
+///Shows the loaded reagents
+/datum/component/chem_booster/proc/show_meds_beaker_contents()
+	if(!meds_beaker.reagents.total_volume)
+		to_chat(wearer, "<span class='notice'>The system's reagent storage is empty.</span>")
+		return
+	to_chat(wearer, "<span class='notice'>The system's reagent storage contains:</span>")
+	for(var/X in meds_beaker.reagents.reagent_list)
+		var/datum/reagent/R = X
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:399]: Use AS

Comment (TiviPlus) [reply code/datums/components/chem_booster.dm:399]: Say to whom it shows

File: code/datums/components/chem_booster.dm [hunk 6]
```diff
@@ -39,6 +40,10 @@
 	var/obj/item/connected_weapon
 	///When was the effect activated. Used to activate negative effects after a certain amount of use
 	var/processing_start = 0
+	///Internal reagent storage used to store and automatically inject reagents into the wearer
+	var/obj/item/reagent_containers/glass/beaker/meds_beaker = null
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:44]: Excess = null

File: code/datums/components/chem_booster.dm [hunk 7]
```diff
@@ -48,13 +53,15 @@
 		return COMPONENT_INCOMPATIBLE
 	update_boost(boost_tier1)
 	analyzer = new
+	meds_beaker = new /obj/item/reagent_containers/glass/beaker
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:56]: No need for the full path

File: code/datums/components/chem_booster.dm [hunk 8]
```diff
@@ -84,6 +91,7 @@
 /datum/component/chem_booster/proc/examine(datum/source, mob/user)
 	SIGNAL_HANDLER
 	to_chat(user, "<span class='notice'>The chemical system currently holds [resource_storage_current]u of green blood. Its' enhancement level is set to [boost_amount+1].</span>")
+	show_meds_beaker_contents()
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:94]: Will it not tochat null if it's on the floor when I observe it?

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/datums/components/chem_booster.dm (hunk 1 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:45]: changed

File: code/datums/components/chem_booster.dm (hunk 2 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:64]: changed

File: code/datums/components/chem_booster.dm (hunk 3 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:354]: changed

File: code/datums/components/chem_booster.dm (hunk 4 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:358]: changed

File: code/datums/components/chem_booster.dm (hunk 5 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:399]: changed

File: code/datums/components/chem_booster.dm (hunk 6 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:44]: changed

File: code/datums/components/chem_booster.dm (hunk 7 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:56]: changed

File: code/datums/components/chem_booster.dm (hunk 8 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm:94]: changed to use the user

File: code/datums/components/chem_booster.dm [hunk 9]
```diff
@@ -330,6 +345,59 @@
...
+	to_chat(wearer, "<span class='notice'>You load [trans]u into the system's reagent storage.</span>")
+	show_meds_beaker_contents(wearer)
+
+///Shows the loaded reagents to the person examining the parent/wearer
+/datum/component/chem_booster/proc/show_meds_beaker_contents(mob/user)
+	if(!meds_beaker.reagents.total_volume)
+		to_chat(user, "<span class='notice'>The system's reagent storage is empty.</span>")
+		return
+	to_chat(user, "<span class='notice'>The system's reagent storage contains:</span>")
+	for(var/datum/reagent/R as() in meds_beaker.reagents.reagent_list)
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:398]:
```suggestion
	for(var/datum/reagent/R AS in meds_beaker.reagents.reagent_list)
```

File: code/datums/components/chem_booster.dm [hunk 10]
```diff
@@ -330,6 +345,59 @@
...
+
+	if(meds_beaker.reagents.total_volume >= meds_beaker.volume)
+		to_chat(wearer, "<span class='notice'>The system's reagent storage is full. You may consider unloading it if you want to load a different mix.</span>")
+		return
+
+	if(!do_after(wearer, 0.5 SECONDS, TRUE, held_item, BUSY_ICON_FRIENDLY, null, PROGRESS_BRASS, ignore_turf_checks = TRUE))
+		return
+
+	var/trans = held_beaker.reagents.trans_to(meds_beaker, held_beaker.amount_per_transfer_from_this)
+	to_chat(wearer, "<span class='notice'>You load [trans]u into the system's reagent storage.</span>")
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:389]:
```suggestion
	to_chat(wearer, "<span class='notice'>You load [trans] units into the system's reagent storage.</span>")
```

**Review by TiviPlus — APPROVED** [review]

File: code/datums/components/chem_booster.dm [hunk 11]
```diff
@@ -199,8 +211,11 @@
 	update_boost(boost_amount*2, FALSE)
 	power_action.action_icon_state = "cboost_on"
 	power_action.update_button_icon()
-	to_chat(wearer, "<span class='notice'>Commensing reagent injection.</span>")
 	playsound(get_turf(wearer), 'sound/effects/bubbles.ogg', 30, 1)
+	to_chat(wearer, "<span class='notice'>Commensing reagent injection.<b>[(automatic_meds_use && meds_beaker.reagents.total_volume) ? " Adding additional reagents." : ""]</b></span>")
+	if(automatic_meds_use)
+		meds_beaker.reagents.trans_to(wearer, 60)
+		scan_user()
```
Comment (TiviPlus) [inline code/datums/components/chem_booster.dm:218]: Does this really need to open a popup when you turn it on

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/datums/components/chem_booster.dm (hunk 11 above)
Comment (Walarks) [reply code/datums/components/chem_booster.dm]: replaced scan with chat message saying what reagents were injected

**Review by TiviPlus — APPROVED** [review]

**Comment by tgstation-server** [comment]
Your Fix/Feature pull request delta is currently below zero (-41). Maintainers may close future Feature/Tweak/Balance PRs. Fixing issues or helping to improve the codebase will raise this score.

**Pull request merged**